src/Interpreters/Context.h

```diff
@@ -157,6 +157,26 @@ struct SharedContextHolder
     std::unique_ptr<ContextShared> shared;
 };
 
+/** PartUuids is a uuid set to control query deduplication.
+ * The object is used in query context in both direction:
+ *  Server->Client to send all parts' UUIDs that have been read during the query
+ *  Client->Server to ignored specified parts from being processed.
+ *
+ *  Current implementation assumes a user setting send_parts_uuid=1 is set.
+ */
+struct PartUuids
```

Comment:
Context is not the best place for this struct. Let's make a separate file in `Storages/MergeTree`.